src/raft/recv_install_snapshot.c

```diff
@@ -328,34 +279,482 @@ static const struct sm_conf leader_states[LS_NR] = {
+		.cs = msg->signature_result.cs,
+		.cs_nr = msg->signature_result.cs_nr,
+	};
+
+	struct raft_io_async_work *work;
+	work = raft_malloc(sizeof(*work));
+	assert(work != NULL);
+	*work = (struct raft_io_async_work) {
+		.work = insert_checksums,
+		.data = data,
+	};
+	state->io->async_work(work, calculate_local_checksums);
+}
+
+void leader_tick(struct sm *leader, const struct raft_message *msg)
+{
+	(void)leader_states;
+
+	assert(leader != NULL);
+	assert(msg != NULL);
+
+	struct snapshot_leader_state *state =
+		CONTAINER_OF(leader, struct snapshot_leader_state, sm);
+	struct snapshot_leader_io *io = state->io;
+	assert(state != NULL && io != NULL);
+
+	PRE(is_main_thread());
+	PRE(msg->server_id == state->follower_id);
+	// TODO: timeouts.
+	int leader_state = sm_state(leader);
+	if (leader_state == LS_FOLLOWER_ONLINE) {
+		if (msg->type != RAFT_IO_APPEND_ENTRIES_RESULT) {
+			return;
+		}
+		raft_index follower_index =
+			msg->append_entries_result.last_log_index;
+		if (!io->log_index_found(follower_index)) {
+			// Follower needs an entry which is not on the Raft log anymore.
+			async_send_install_snapshot(state, msg);
+		} else {
+			sm_move(leader, LS_FOLLOWER_ONLINE);
+		}
+	} else if (leader_state == LS_FOLLOWER_NEEDS_SNAPSHOT) {
+		if (msg->type != RAFT_IO_INSTALL_SNAPSHOT_RESULT) {
+			return;
+		}
+		// TODO: control that we do not move from stale state on async callback.
+		// what happens if we finish the async callback after we received another message.
+		// create type state_transition_t.
+		PRE(state->ht == NULL && state->ht_stmt == NULL);
+
+		// TODO: sem_move. Create a mutex to sync.
+		struct raft_message *reply = get_signature_message(state, msg);
+		async_create_ht_and_reply(state, reply, LS_SIGNATURES_CALC_STARTED);
+	} else if (leader_state == LS_SIGNATURES_CALC_STARTED) {
+		if (msg->type != RAFT_IO_SIGNATURE_RESULT) {
+			return;
+		}
+		PRE(state->ht != NULL && state->ht_stmt != NULL);
+
+
+		if (msg->signature_result.result == DONE) {
+			async_insert_checksums_calculate_local(state, msg);
+		} else if (msg->signature_result.result == OK) {
+			struct raft_message *reply = get_signature_message(state, msg);
+			async_insert_checksums_send_reply(state, msg, reply, LS_SIGNATURES_CALC_STARTED);
+		}
+	} else if (leader_state == LS_SNAPSHOT_INSTALLATION_STARTED) {
+		/* We are not expecting any message. */
+	} else if (leader_state == LS_SNAPSHOT_CHUNCK_SENT) {
+		if (msg->type != RAFT_IO_INSTALL_SNAPSHOT_MV_RESULT &&
+				msg->type != RAFT_IO_INSTALL_SNAPSHOT_CP_RESULT) {
+			return;
+		}
+		process_snapshot_installation_cp_or_mv(state, msg);
+		if (state->last_page_acked >= state->last_page) {
+			sqlite3_finalize(state->ht_stmt);
+			/* No more pages to sent, we are done. */
+			async_send_install_snapshot_done(state, msg);
+			return;
+		}
+		async_send_mv_or_cp(state);
+	}
+}
+
+__attribute__((unused)) static bool leader_invariant(const struct sm *sm,
+						     int prev_state)
+{
+	bool rv;
```

Comment:
I would prefer that `rv` always be an `int` (that is, to use a different name here).